geopandas/clip.py

```diff
@@ -0,0 +1,228 @@
+"""
+geopandas.clip
+============
+
+A module to clip vector data using GeoPandas.
+
+"""
+
+
+import pandas as pd
+import geopandas as gpd
+
+
+def _clip_points(shp, clip_obj):
+    """Clip point geometry to the clip_obj GeoDataFrame extent.
+
+    Clip an input point GeoDataFrame to the polygon extent of the clip_obj
+    parameter. Points that intersect the clip_obj geometry are extracted with
+    associated attributes and returned.
+
+    Parameters
+    ----------
+    shp : GeoDataFrame
+        Composed of point geometry that is clipped to clip_obj.
+
+    clip_obj : GeoDataFrame
+        Reference polygon for clipping.
+
+    Returns
+    -------
+    GeoDataFrame
+        The returned GeoDataFrame is a subset of shp that intersects
+        with clip_obj.
+    """
+    poly = clip_obj.geometry.unary_union
+    return shp[shp.geometry.intersects(poly)]
+
+
+def _clip_multi_point(shp, clip_obj):
```

Comment:
To understand why you are doing the "multi" separately with the additional explode/dissolve step (as in principle intersection on MultiPoints should work fine as well?) 
Is it be able to use `intersects` instead of `intersection`? (which might be faster?)

Reply:
good question @jorisvandenbossche , i think we went this route because we couldn't think of an edge case where points would  return slivers (lines or polygons) and thus the code to clip points is simpler.

With that said i also noticed @martinfleis note above that we aren't using rtree for points which could be a bit slower. We didn't test it as much because points seem to clip quickly (we didn't test this on millions of points tho!).

Would you prefer the helper functions to be combined with a conditional for points? it just seemed a bit cleaner to keep them separate from a maintenance perspective given the added complexity that polygons and lines bring. 

i'm really open to suggestions. i do think this code could be optimized more and parallel would be really great. but that could be a future enhancement too.